topdown/numbers.go

```diff
@@ -88,3 +102,33 @@ func init() {
 	RegisterBuiltinFunc(ast.NumbersRange.Name, builtinNumbersRange)
 	RegisterBuiltinFunc(ast.RandIntn.Name, builtinRandIntn)
 }
+
+type randGeneratorKeyType string
+
+var randGeneratorKey = randGeneratorKeyType("random")
```

Comment:
Hmm. If we plan to only have one, we could use a
```go
type randGeneratorKey struct{}
```
instead, I think. But it doesn't hurt. (I mean, if there's only ever one, we could also use a sync.Once. It's all the same in the end.)

Was a bit surprised to see this getting added to the cache instead of being a field on BuiltinContext. But it's fine either way, I suppose.

Another thing, is there something indeterminate with multiple readers of `bctx.Seed`? I mean this to getRandomNumberGenerator could be one user of bctx.Seed among many, and I can't tell from the top of my head if that's fine or not.

Reply:
Making it an empty struct would work.

Was a bit surprised to see this getting added to the cache instead of being a field on BuiltinContext. But it's fine either way, I suppose.

I put the generator into the cache because it didn't feel right to have the built-in implementation lazily initializing fields on the BuiltinContext struct. Having said that, we could just move the getRandomNumberGenerator function onto the BuiltinContext struct and use that.

Another thing, is there something indeterminate with multiple readers of bctx.Seed? I mean this to getRandomNumberGenerator could be one user of bctx.Seed among many, and I can't tell from the top of my head if that's fine or not.

It'll be deterministic as long as calls to read from the bctx.Seed occur in the same order. Admittedly, this is the bare minimum we can do for ensuring query results are reproducible. It'll be harder to guarantee across different versions of OPA if expressions end up evaluated in a different order. Longer term, I think we want a better solution that has the evaluator record outputs from impure functions. This would let us replay evaluation by replacing the function calls with cache lookups.

Reply:
Thanks, that's helping. I'm fine either way, with a slight preference for the empty struct because it conveys "there's only one" in a different way. But this is splitting hairs, let's get this merged.

Reply:
I've moved the function over to the BuiltinContext struct and used a private member to store the RNG. PTAL though I'm happy w/ the latest version (I think).